### PR TITLE
tuntox: init at 0.0.8

### PR DESCRIPTION
**Pull request by fgaz**
###### Motivation for this change

Add the [tuntox](https://github.com/gjedeer/tuntox) package and a systemd service to set up a server with it. I may also add a tuntox client service in the future, so the server is under the `tuntox.server` namespace

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
  - **This causes the PreStart script to fail!** Any idea why? Maybe I shouldn't create that directory that way, but I couldn't find alternatives.. or maybe `/var` isn't the right place?
    ```
    tuntox.service: Failed at step NAMESPACE spawning /nix/store/wa0c5wzs92ngv7qbsbi0c3dvlxwk2fs0-unit-script/bin/tuntox-pre-start: No such file or directory
    ```
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
@GrahamcOfBorg eval

**Comment by orivej** [comment]
> This causes the PreStart script to fail!

What causes the PreStart script to fail? How can I reproduce it?

**Comment by fgaz** [comment]
@orivej useSandbox prevents the preStart script from creating/modifying/(accessing?) /var

Just enable that option and set `tuntox.server.enable = true` and it should give that error in the journal (`journalctl -b -u tuntox`)

File: nixos/modules/misc/ids.nix
Comment (ryantm) [inline nixos/modules/misc/ids.nix]: Are you sure you need the user to be a specific ID? It usually isn't necessary since the system can assign user IDs.

File: nixos/modules/services/networking/tuntox.nix
Comment (ryantm) [inline nixos/modules/services/networking/tuntox.nix]: Instead of making the directory in a preStart you can do it in an activation script like in https://github.com/NixOS/nixpkgs/blob/b43c4d8b758a3965067b24f3796cdd532756a2ce/nixos/modules/services/web-apps/mattermost.nix#L165

File: pkgs/tools/networking/tuntox/default.nix
Comment (ryantm) [inline pkgs/tools/networking/tuntox/default.nix]: Can just be `rev = version;`.

File: nixos/modules/services/networking/tuntox.nix
Comment (ryantm) [inline nixos/modules/services/networking/tuntox.nix]: Please use `mkEnableOption` here.

File: nixos/modules/services/networking/tuntox.nix
Comment (ryantm) [inline nixos/modules/services/networking/tuntox.nix]: It might be good to point out that this secret will be world readable when inserted in the Nix store. Can this optionally be done with a secretFile that is stored outside the nix store?

**Review by ryantm — CHANGES_REQUESTED** [review]

**Comment by Ekleog** [comment]
(triage) @fgaz Are you still willing to work on this?

**Comment by fgaz** [comment]
@Ekleog yes, but it'll have to wait a bit

**Comment by aanderse** [comment]
@fgaz do you need a hand with this or are you able to continue?

**Comment by fgaz** [comment]
@aanderse I just don't (and won't for another month) have much time to make and test those modifications. Maybe in the meantime I could split this into one pr for the package, which is quick and already reviewed, and keep this one for the module only.

File: pkgs/tools/networking/tuntox/default.nix
Comment (aanderse) [inline pkgs/tools/networking/tuntox/default.nix]: Please replace with `pname = "tuntox";` and then you can drop `name` as it will be implied.

File: pkgs/tools/networking/tuntox/default.nix
Comment (aanderse) [inline pkgs/tools/networking/tuntox/default.nix]: `repo = pname;` is a common pattern.

File: pkgs/tools/networking/tuntox/default.nix
Comment (aanderse) [inline pkgs/tools/networking/tuntox/default.nix]: Please quote.

**Review by aanderse — CHANGES_REQUESTED** [review]
@fgaz Splitting this PR into 2 sounds like a good idea if you don't have the time but want to get the package in ASAP. If you're not in a rush, though, no worries because neither are we :smile:

**Comment by ivan** [comment]
For the author, reviewers, and committers: this PR was scanned and appears to add a use of the deprecated `types.string`, which emits a warning as of #66346. Before merging, please change this to another type, possibly:

- `types.str` for a single string where merging does not make sense, or cannot work
- `types.lines` for multi-line configuration or scripts where merging is possible
- `types.listOf types.str` for a mergeable list of strings

**Comment by fgaz** [comment]
Stale bot, I'll still finish this... someday

**Comment by willcohen** [comment]
@fgaz Was looking at putting together a PR for tuntox and found this draft first. Are you still interested in finishing it? I'd also be happy to start a new PR based off of this WIP branch. I've never had to create a package that creates/enables a service there so I'm a little bit less certain about fine-tuning those details, but at the very least would like to see the binaries available!

**Pull request closed without merging**
